src/core/server/http/base_path_service.test.ts

```
@@ -0,0 +1,132 @@
+      expect(basePath.get(request)).toBe('/baz/');
+    });
+
+    it('returns base path associated with an incoming KibanaRequest', () => {
+      const request = httpServerMock.createRawRequest();
+      const basePath = new BasePath();
+
+      basePath.set(KibanaRequest.from(request, undefined), '/baz/');
+      expect(basePath.get(KibanaRequest.from(request, undefined))).toBe('/baz/');
+    });
+
+    it('operates with both Legacy.Request/KibanaRequest requests', () => {
+      const request = httpServerMock.createRawRequest();
+      const basePath = new BasePath();
+
+      basePath.set(request, '/baz/');
+      expect(basePath.get(KibanaRequest.from(request, undefined))).toBe('/baz/');
+    });
+
+    it('is based on server base path', () => {
+      const request = httpServerMock.createRawRequest();
+      const basePath = new BasePath('/foo/bar');
+
+      basePath.set(request, '/baz/');
+      expect(basePath.get(request)).toBe('/foo/bar/baz/');
+    });
+  });
+
+  describe('#set()', () => {
+    it('#set() cannot be set twice for one request', () => {
+      const request = httpServerMock.createRawRequest();
+      const basePath = new BasePath('/foo/bar');
+
+      const setPath = () => basePath.set(request, 'baz/');
+      setPath();
+
+      expect(setPath).toThrowErrorMatchingInlineSnapshot(
+        `"Request basePath was previously set. Setting multiple times is not supported."`
+      );
+    });
+  });
+
+  describe('#prepend()', () => {
+    it('adds the base path to the path if it is relative and starts with a slash', () => {
+      const basePath = new BasePath('/foo/bar');
+
+      expect(basePath.prepend('/a/b')).toBe('/foo/bar/a/b');
+    });
+
+    it('leaves the query string and hash of path unchanged', () => {
+      const basePath = new BasePath('/foo/bar');
+
+      expect(basePath.prepend('/a/b?x=y#c/d/e')).toBe('/foo/bar/a/b?x=y#c/d/e');
+    });
+
+    it('returns the path unchanged if it does not start with a slash', () => {
```

Optional: This is the same as the existing behaviour, but I'm not sure why we implemented this behaviour. It feels like your code can silently fail if you forget a `/` in a path you're prepending. The only way you would notice this mistake is by testing it and noticing that the basepath was never added which feels brittle.

Is there a reason we can't handle this case by inserting a `/`? If there is a good reason we don't want to accept strings not starting with `/` should we perhaps throw an exception to show that this is unexpected/unsupported behaviour instead of pretending like we're giving you a prepended basepath when we are not.

---

ping @spalger as an author

---

I agree that the implicitness is undesirable. I originally did it this way so that the code would ignore fully resolved urls, but I'm not sure if that's being used... sounds like a really tricky thing to figure out.